Implement behaviour of yield* as specced by rev32 of the ES6 draft.

mnieper opened the pull request:
This should close #1688.

arv commented:
You should probably rebase with upstream/master. Your PR has a lot of commits that have been committed already.

arv commented on src/runtime/generators.js:
```diff
@@ -130,14 +131,19 @@ GeneratorContext.prototype = {
           throw e;
-        }
-        if (iterator.throw) {
+        } else if (iterator.throw) {
```
no else after return/throw

arv commented on test/feature/Yield/BadIterable.js:
```diff
@@ -0,0 +1,44 @@
+var i1 = new BadIterable();
+var g1 = wrap(i1);
+assert.deepEqual(g1.next(), {value: 42, done: false});
+assert.throws(() => g1.throw("ex1"), TypeError);
```
we use single quotes for consistency

arv commented on test/feature/Yield/BadIterable.js:
```diff
@@ -0,0 +1,44 @@
+var g1 = wrap(i1);
+assert.deepEqual(g1.next(), {value: 42, done: false});
+assert.throws(() => g1.throw("ex1"), TypeError);
+assert.equal(i1.closed, true);
```
you can do isTrue here

arv commented:
LGTM

Could you squash these into one commit?

mnieper commented:
Did the corrections. Have to read about squashing and rebasing. Don't want to wreck my repository as I have a number of branches open.

arv commented:
If you don't want to rebase/squash I can do that before merging to master.

mnieper commented:
Sorry, too late here already. Thus the commits are not yet squashed (my first attempt failed) but the upstream master branch has been merged into.

Do you happen to know a quick guide for the squashing and rebasing?

johnjbarton commented:
When we say "rebase" we mean the equivalent of:
```
git checkout master
git pull --ff-only upstream master  # origin master syncs to upstream.
git checkout <branch>
git rebase master  # branch syncs to master
```
This affects just two branches, master and `<branch>`.  

When we say squash, its like this:
http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

See also http://eagain.net/articles/git-for-computer-scientists/

HTH

UltCombo commented:
You can skip pulling the master by rebasing directly on upstream/master btw:

```
# in case you haven't added the upstream remote yet:
git remote add upstream https://github.com/google/traceur-compiler.git
git fetch --all
# on your "yield-star" branch
git rebase upstream/master
```

UltCombo commented:
By the way, if you're going to try to squash commits for the first time, create another branch from your current branch first:

```
git checkout -b test
```

Because odds that something will go terribly wrong are quite high (talking from own experience), so if the worst happens you can just `git checkout yield-star` and create another branch to try again.

mnieper commented:
Rebased and squashed, although the result does not look perfect. :-(

@arv So please do any further squashing that needs to be done; should be ready for merge.

arv commented:
Committed as 28baf5195da2905b5ff3fc2a6456dd0c86e7e514

I ended up reconstructing the commit.

```sh
git reset master
git add ...
git commit --author="..."
```

mnieper commented:
Thanks. Closing this.
